finality-aleph/src/sync/handler/mod.rs

```diff
@@ -559,9 +569,175 @@ mod tests {
+        assert!(
+            handler.interest_provider().get(top)
+                == Interest::Required {
+                    know_most,
+                    branch_knowledge: LowestId(bottom.clone()),
+                },
+            "should require the top"
+        );
+    }
+
+    fn grow_light_branch(
+        handler: &mut Handler<
+            MockBlock,
+            MockPeerId,
+            MockJustification,
+            Backend,
+            Backend,
+            Backend,
+            Backend,
+        >,
+        bottom: &MockIdentifier,
+        length: usize,
+        peer_id: MockPeerId,
+    ) -> Vec<MockHeader> {
+        let branch: Vec<_> = bottom.random_branch().take(length).collect();
+        let top = branch.last().expect("branch should not be empty").id();
+
+        assert!(
+            handler.handle_internal_request(&top).expect("should work"),
+            "should be newly required"
+        );
+        assert!(
+            handler.interest_provider().get(&top)
+                == Interest::Required {
+                    know_most: HashSet::new(),
+                    branch_knowledge: LowestId(top.clone()),
+                },
+            "should be required"
+        );
+
+        let (maybe_id, maybe_error) = handler.handle_request_response(
+            branch
+                .iter()
+                .cloned()
+                .rev()
+                .map(ResponseItem::Header)
+                .collect(),
+            peer_id,
+        );
+
+        assert!(
+            maybe_id.is_none(),
+            "should not create new highest justified"
+        );
+        assert!(maybe_error.is_none(), "should work");
+
+        branch
+    }
+
+    fn create_dangling_branch(
+        handler: &mut Handler<
+            MockBlock,
+            MockPeerId,
+            MockJustification,
+            Backend,
+            Backend,
+            Backend,
+            Backend,
+        >,
+        height: BlockNumber,
+        length: usize,
+        peer_id: MockPeerId,
+    ) -> (MockIdentifier, MockIdentifier) {
+        let bottom = MockIdentifier::new_random(height);
+        let top = grow_light_branch(handler, &bottom, length, peer_id)
+            .last()
+            .expect("branch should not be empty")
+            .id();
+        assert_dangling_branch_required(handler, &top, &bottom, HashSet::from_iter(vec![peer_id]));
+        (bottom, top)
+    }
+
+    fn finalize_light_branch(
+        handler: &mut Handler<
+            MockBlock,
+            MockPeerId,
+            MockJustification,
+            Backend,
+            Backend,
+            Backend,
+            Backend,
+        >,
+        branch: Vec<MockHeader>,
+        peer_id: MockPeerId,
+    ) {
+        let blocks: Vec<MockBlock> = branch
+            .iter()
+            .cloned()
+            .map(|header| MockBlock::new(header, true))
+            .collect();
+        let (maybe_id, maybe_error) = handler.handle_request_response(
```

Comment:
can we use handle_justification here or rename the fn

Reply:
Which function do you want to rename? I need to import a list of blocks and justifications here, this seems like a proper approach to me.

Reply:
the function is called `finalize_light_branch` but potentially we are actually doing a lot more here by using handle_request_response. Now that I think about it ideally, we should call the backend directly.

Reply:
ok, I get it

the thing is that I'd like to have such small primitives for building larger tests, but these also should use `Handler` public api as much as possible

calling backend directly, and then using some `handle_justification_SOMEHOW` method would also finalize the branch, but it rather simulates the case where the Substrate code syncs and imports blocks, and then we get the justifications SOMEHOW

I'm not sure what to do in such situations yet - should I write multiple `finalize_light_branch_*` functions, and then choose randomly which one to use in the other tests? or choose a particular one, and test the others in separate unit tests just once? do you have any ideas?

Reply:
hmm in such case I would rename this fn to handle_incoming_justifications_for_branch, wdyt?

Reply:
I removed the function `finalize_light_branch` and added a function that just prepares the response:
```
fn finalizing_response(
    branch: Vec<MockHeader>,
    include_headers: bool,
    include_blocks: bool,
    include_justifications: bool,
) -> MockResponseItems { ... }
```
I do hope that this will all clarify in the next PRs.